packages/replayio/src/utils/recordings/upload/uploadRecording.ts

```diff
@@ -198,54 +194,115 @@ async function uploadRecordingFileInParts({
   partLinks: string[];
   recordingPath: string;
 }): Promise<string[]> {
-  const { size: totalSize } = statSync(recordingPath);
-
-  const results = await promiseMap<string, string>(
-    partLinks,
-    async (url: string, index: number) => {
-      return retryWithLinearBackoff(
-        async () => {
-          const partNumber = index + 1;
-          const start = index * chunkSize;
-          const end = Math.min(start + chunkSize, totalSize) - 1; // -1 because end is inclusive
-
-          debug("Uploading part %o", {
-            partNumber,
-            start,
-            end,
-            totalSize,
-            chunkSize,
-          });
-          return uploadPart(url, { recordingPath, start, end }, end - start + 1);
-        },
-        error => {
-          debug(`Failed to upload part ${index + 1}. Will be retried: %o`, error);
-        },
-        10
-      );
-    },
-    { concurrency: 10 }
-  );
+  const { size: totalSize } = await stat(recordingPath);
+  const abortController = new AbortController();
+
+  const partsUploads = partLinks.map((url: string, index: number) => {
+    return uploadQueue.add(async () => {
+      try {
+        return retryWithLinearBackoff(
+          async () => {
+            const partNumber = index + 1;
+            const start = index * chunkSize;
+            const end = Math.min(start + chunkSize, totalSize) - 1; // -1 because end is inclusive
+
+            debug("Uploading part %o", {
+              partNumber,
+              start,
+              end,
+              totalSize,
+              chunkSize,
+            });
+            return uploadPart(
+              { url, recordingPath, start, end, size: end - start + 1 },
+              abortController.signal
+            );
+          },
+          (error: any) => {
+            debug(`Failed to upload part ${index + 1}. Will be retried: %o`, error);
+            if (error.code === "ENOENT") {
+              throw error;
+            }
+          }
+        );
+      } catch (error) {
+        abortController.abort();
```

Comment:
There is probably no point in uploading other parts if one of them failed to be uploaded (note that they are retried before this handler kicks in). So I decided to add some cancellation to all of this.